Moving subquery execution to a worker thread

arsen-es opened the pull request:
*Issue #, if available:* N/A

*Description of changes:* Run queries including a subquery on a worker thread. This change does not affect the execution of other types of queries. This is just a first step to unblock the queries with `WHERE field IN (subquery)` syntax, but still needs improvement including but not limited to the following:
 1. the condition of whether the query has a subquery should ideally be derived from the Sql parser, not by string manipulation,
 1. the explain call should not execute the subquery, but should just explain how it will be executed.

*How tested:* All integ tests pass, manually tested on queries that failed previously (with jvm assertions enabled) and they pass with this change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dai-chen commented:
IMHO, trying to avoid query execution in explain is the better way (or at least in this case of subquery). The code would be clean, same logic for different explain and no need for hack like in your `containsSubquery`. Unless customer really wants an executable DSL from explain anyway. I think explain of Join, especially Nested Loop, could be your reference.

arsen-es commented:
@dai-chen The intention was to unblock this use case before coming up with the more structured and generic solution, so I agree that this is not the best way. Considering we don't claim supporting subqueries yet, I am closing this PR, and will work on the better solution.

dai-chen commented:
> @dai-chen The intention was to unblock this use case before coming up with the more structured and generic solution, so I agree that this is not the best way. Considering we don't claim supporting subqueries yet, I am closing this PR, and will work on the better solution.

Yeah, my concern is we're introducing a major code change which is very likely to refactor or even revert later. Thanks for your effort on this anyway!

penghuo commented on src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java:
```diff
@@ -147,16 +205,16 @@ private boolean isClientError(Exception e) {
 
-    private RestChannelConsumer reportError(Exception e, RestStatus status) {
-        return sendResponse(new ErrorMessage(e, status.getStatus()).toString(), status);
+    private void sendResponse(final RestChannel channel, final String message, final RestStatus status) {
```
is it possible to avoid these unnecessary change.

arsen-es replied:
the return type is changed, this is necessary.

arsen-es commented:
Re-opening, as we are considering having a quick fix before we roll out the better support for sub-queries, which might take longer time.

penghuo commented on src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java:
```diff
@@ -136,7 +140,61 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
+
+    private static boolean containsSubquery(final SqlRequest request) {
+
+        return request.getSql().toLowerCase(Locale.ROOT).contains("(select ");
```
how to handle "( select"? Is there any preprocessing already handle this?

arsen-es replied:
I can check for regex match, instead of hard-coded string, e.g.:
```
return request.getSql().toLowerCase(Locale.ROOT).matches("\\(\\s*select ");
```

arsen-es commented:
> 
> 
> Becase this PR is the quick fix for subquery, could you make it easy to revert in the future?
> e.g. remove unnecessary code change.

Seems like I can revert the change of (need to double check)
src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequestFactory.java

dai-chen commented:
> > Becase this PR is the quick fix for subquery, could you make it easy to revert in the future?
> > e.g. remove unnecessary code change.
> 
> Seems like I can revert the change of (need to double check)
> src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequestFactory.java

And it's saying "This branch has conflicts that must be resolved". Could you merge latest changes in master too in your new revision?

arsen-es commented:
> 
> 
> > > Becase this PR is the quick fix for subquery, could you make it easy to revert in the future?
> > > e.g. remove unnecessary code change.
> > 
> > 
> > Seems like I can revert the change of (need to double check)
> > src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequestFactory.java
> 
> And it's saying "This branch has conflicts that must be resolved". Could you merge latest changes in master too in your new revision?

Sure @dai-chen, I should have done that in the first place. Working on it now...

arsen-es commented:
I have addressed all the comments in #115 . Couldn't update this PR with the changes, since it wanted me to rebase the branch, and that would include too many commits, that's why moved to a new PR. Please continue the review there, I am closing this one.
